openApiSync/openApiSync.go

```diff
@@ -0,0 +1,205 @@
+			}
+		} else {
+			pass.Reportf(f.Comment.Pos(), "Router field %s in Router struct is not commented properly\n", routerName)
+		}
+	}
+}
+
+func parseInitFunction(pass *analysis.Pass, decl *ast.FuncDecl, routerPrefixes map[string]string, initFunctions []string) []string {
+	for _, stmt := range decl.Body.List {
+		if expr, ok := stmt.(*ast.ExprStmt); ok {
+			if call, ok := expr.X.(*ast.CallExpr); ok {
+				sel := call.Fun.(*ast.SelectorExpr)
+				if sel.X.(*ast.Ident).Name == "api" {
+					initFunctions = append(initFunctions, sel.Sel.Name)
+				}
+			}
+		} else if assgn, ok := stmt.(*ast.AssignStmt); ok {
+			if len(assgn.Lhs) == 1 && strings.HasPrefix(formatNode(pass.Fset, assgn.Lhs[0]), "api.BaseRoutes") {
+				subRouterName := formatNode(pass.Fset, assgn.Lhs[0])[15:]
+				if subRouterName == "ApiRoot" || subRouterName == "Root" {
+					continue
+				}
+				rhs := formatNode(pass.Fset, assgn.Rhs[0])[15:]
+				router := rhs[:strings.Index(rhs, ".")]
+				path := rhs[strings.Index(rhs, ".")+13 : strings.LastIndex(rhs, ".")-2]
+				prefix := ""
+				switch router {
+				case "ApiRoot":
+					prefix = "api/v4"
+				case "Root":
+					prefix = ""
+				default:
+					if s, ok := routerPrefixes[router]; ok {
+						prefix = s
+					} else {
+						pass.Reportf(assgn.Rhs[0].Pos(), "cannot find prefix for %s\n", router)
+
+					}
+				}
+				s := fmt.Sprintf("%v%v", prefix, path)
+				s2 := routerPrefixes[subRouterName]
+				if s2 != s {
+					pass.Reportf(assgn.Rhs[0].Pos(), "PathPrefix doesn't match field comment for field '%s': '%s' vs '%s'\n", subRouterName, s, s2)
+				}
+
+			}
+		}
+	}
+	return initFunctions
+}
+
+func validateComments(pass *analysis.Pass) ([]string, map[string]string) {
+	initFunctions := []string{}
+	routerPrefixes := map[string]string{}
+	for _, file := range pass.Files {
+		ast.Inspect(file, func(n ast.Node) bool {
+			if decl, ok := n.(*ast.GenDecl); ok && len(decl.Specs) == 1 {
+				parseRoutesStruct(pass, decl, routerPrefixes)
+			} else if decl, ok := n.(*ast.FuncDecl); ok && decl.Name.Name == "Init" {
+				initFunctions = parseInitFunction(pass, decl, routerPrefixes, initFunctions)
+			}
+			return true
+		})
+	}
+	return initFunctions, routerPrefixes
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	if specFile == "" {
+		return nil, errors.New("Please supply a path to OpenAPI spec yaml file via -openApiSync.spec")
+	}
+	if _, err := os.Stat(specFile); err != nil {
+		return nil, errors.Wrapf(err, "spec file does not exist")
+	}
+	swagger, err := openapi3.NewSwaggerLoader().LoadSwaggerFromFile(specFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Unable to parse swagger")
+	}
+
+	initFunctions, routerPrefixes := validateComments(pass)
+
+	var swaggerPaths []string
+	for p := range swagger.Paths {
+		swaggerPaths = append(swaggerPaths, p)
+	}
+	model := fuzzy.NewModel()
+	model.Train(swaggerPaths)
+
+	for _, initFunc := range initFunctions {
```

Comment:
This loop looks really unnefficient, you scanning the whole AST for all the files for every single function, you can extract that loop here, process the AST only one time, and execute the logic around the functions in this same pass.